src/org/zaproxy/zap/extension/ascanrules/TestCrossSiteScriptV2.java

```diff
@@ -355,24 +373,36 @@ public void scan(HttpMessage msg, String param, String value) {
         	            	// Maybe they're blocking script tags
             	            contexts2 = performAttack (msg, param, 
     			            		"<b onMouseOver=alert(1);>test</b>", context, HtmlContext.IGNORE_PARENT);
-                            if (contexts2 == null) {
-                                break;
-                            }
-    			            for (HtmlContext context2 : contexts2) {
-    			            	if ("body".equalsIgnoreCase(context2.getParentTag()) ||
-    			            			"b".equalsIgnoreCase(context2.getParentTag())) {
-    			            		// Yep, its vulnerable
-    								bingo(Alert.RISK_HIGH, Alert.CONFIDENCE_MEDIUM, null, param, contexts2.get(0).getTarget(), 
-    										"", contexts2.get(0).getTarget(), contexts2.get(0).getMsg());
-    								attackWorked = true;
-    								break;
-    			            	}
+            	            if (contexts2 != null) {
+	    			            for (HtmlContext context2 : contexts2) {
+	    			            	if ("body".equalsIgnoreCase(context2.getParentTag()) ||
+	    			            			"b".equalsIgnoreCase(context2.getParentTag())) {
+	    			            		// Yep, its vulnerable
+	    								bingo(Alert.RISK_HIGH, Alert.CONFIDENCE_MEDIUM, null, param, contexts2.get(0).getTarget(), 
+	    										"", contexts2.get(0).getTarget(), contexts2.get(0).getMsg());
+	    								attackWorked = true;
+	    								break;
+	    			            	}
+	    			            }
     			            }
+            	            if (!attackWorked) {
+                                // Try double encoded
+                                List<HtmlContext> contexts3 = performAttack (msg, param, 
+                                        URLEncoder.encode(GENERIC_SCRIPT_ALERT, UTF8), 

```

Comment:
My test case was a GET request (sadly post wasn't accepted). In a GET scenario the double encoding happened naturally after a forced initial encoding. I "think" this will work for posting too, but I'm not 100% sure.

Should I check if the variant is other than GET or POST and not perform this attack?

Reply:
I've added a check `if (Arrays.asList(NameValuePair.TYPE_QUERY_STRING, NameValuePair.TYPE_POST_DATA).contains(paramType))`

Reply:
Sounds good, this test does not seem to apply to other variants.

Only the query is automatically re-encoded. You need to encode for POST.